python/src/wslink/backends/aiohttp/relay.py

```diff
@@ -0,0 +1,385 @@
+                await self._ws.pong(msg.data)
+            elif msg.type == aiohttp.WSMsgType.CLOSE:
+                await self._ws.close()
+            else:
+                print("Invalid message to forward")
+        else:
+            print("server::send - NO SEND")
+            print(f"{self._connected} - {self._ws.closed}")
+            print("-" * 60)
+
+    def disconnect(self):
+        print(f"server::disconnect", self._connected)
+        if self._connected > 0:
+            self._connected = -1
+            if self._destination:
+                self._destination.disconnect()
+
+
+# -----------------------------------------------------------------------------
+
+
+class ForwardConnection:
+    def __init__(self, request, url):
+        self._req = request
+        self._url = url
+        self._ws_client = WsClientConnection()
+        self._ws_server = WsServerConnection()
+        self._ws_server.bind(self._ws_client)
+        self._ws_client.bind(self._ws_server)
+
+    async def connect(self):
+        task = asyncio.create_task(self._ws_client.connect(self._url))
+        task.add_done_callback(lambda *args, **kwargs: self._ws_server.disconnect())
+        await self._ws_client.ready
+        await self._ws_server.connect(self._req)
+
+    def disconnect(self):
+        self._ws_client.disconnect()
+        self._ws_server.disconnect()
+
+
+class SinkConnection:
+    def __init__(self, request):
+        self._process_req = request
+        self._client_req = None
+        self._process_ws = None
+        self._client_ws = None
+
+    def can_handle(self, request):
+        if self._process_req == request:
+            return True
+
+        if self._client_req == request:
+            return True
+
+        if self._client_ws is None:
+            return True
+
+        return False
+
+    async def connect(self, request):
+        if self._process_req == request:
+            self._process_ws = WsServerConnection()
+            await self._process_ws.connect(request)
+        elif self._client_req is None:
+            self._client_req = request
+            self._client_ws = WsServerConnection()
+            self._client_ws.bind(self._process_ws)
+            self._process_ws.bind(self._client_ws)
+            await self._client_ws.connect(request)
+
+
+# -----------------------------------------------------------------------------
+# Handlers
+# -----------------------------------------------------------------------------
+
+
+async def _root_handler(request):
+    if request.query_string:
+        return web.HTTPFound(f"index.html?{request.query_string}")
+    return web.HTTPFound("index.html")
+
+
+# -----------------------------------------------------------------------------
+
+
+class WsHandler:
+    def __init__(self):
+        self._forward_map = {}
+        self._relay_map = {}
+
+    def get_handler(self, mode="forward"):
+        logger.info("get_handler", mode)
```

Comment:
A couple of things about the logger:
1. We need to put a call to `logging.basicConfig()` somewhere, which should be called once in the program before any loggers are created (probably just put it in the `main()` function). This is necessary for the `info` and `debug` logging messages to work (since, otherwise, the loggers don't get initialized properly).
2. The logger doesn't work like print statements where you can do things like `logger.info("get_handler", mode)`. Instead, we have to do either `logger.info("get_handler %s", mode)` or `logger.info(f"get_handler {mode}")`.

I'm gonna fix these things, and then the logger should work.

Reply:
After these changes, starting up the relay looks like this:
```
INFO:wslink.backends.aiohttp.relay:Proxy route: /proxy/{port}
INFO:wslink.backends.aiohttp.relay:get_handler forward
INFO:wslink.backends.aiohttp.relay:WWW: ./www
INFO:wslink.backends.aiohttp.relay:Starting relay server: localhost 8080
```